File: rakugan/demo/tasks.pl
# Work that takes a while, done off the window's thread. `task` runs
# the sub on a thread of perl's own; when it answers, `on_done` is
# called on the window's thread with the answer.
#
# Nothing inside the work touches the app's state or the screen. That
# is the whole rule, and it is why the answer comes back as a value
# rather than the work writing it anywhere.
use Rakugan;

class Jobs {
    use Rakugan;
    field $status = "idle";
    field $answer = 0;
    field $done   = 0;

    method start {
        $status = "working";
        task(sub {
            # deliberately slow, and deliberately arithmetic: both runs
            # have to agree about what it answers.
            my $total = 0;
            my $i = 0;
            while ($i < 300000) {
                $total += $i % 7;
                $i += 1;
            }
            $total;
        }, on_done => sub ($v) {
            $answer = $v;
            $done += 1;
            $status = "done";
        });
    }

    method view {
        return column(
            text("background work", size => 18, bold => true),
            text("status: $status"),
            text("answer: $answer  ($done finished)"),
            button("start slow work", on_click => sub { $self->start }),
            spacing => 10,
            padding => 14,
        );
    }
}

run(Jobs->new, title => "tasks");

File: gomamochi/demo/tasks.go
// Work that takes a while, done off the window's thread. `Task` runs
// the work on a goroutine of its own; when it answers, the second
// closure is called on the window's thread with the answer.
//
// Nothing inside the work touches the app's fields or the screen. That
// is the whole rule, and it is why the answer comes back as an argument
// rather than the worker writing it anywhere.
package main

import (
	"fmt"

	. "github.com/i2y/yokan/gomamochi"
)

type Jobs struct {
	status string
	answer int
	done   int
}

func (j *Jobs) start() {
	j.status = "working"
	Task(func() any {
		// deliberately slow, and deliberately arithmetic: both runs
		// have to agree about what it answers.
		total := 0
		for i := 0; i < 300000; i++ {
			total += i % 7
		}
		return total
	}, func(v any) {
		j.answer = v.(int)
		j.done += 1
		j.status = "done"
	})
}

func (j *Jobs) View() Element {
	return Column(
		Text("background work").Size(18).Bold(true),
		Text("status: "+j.status),
		Text(fmt.Sprintf("answer: %d  (%d finished)", j.answer, j.done)),
		Button("start slow work").OnClick(func() { j.start() }),
	).Spacing(10).Padding(14)
}

func main() {
	Run(&Jobs{status: "idle"}, Title("tasks"))
}
